src/ServiceControl/HeartbeatMonitoring/RegisterEndpointStartupHandler.cs

```diff
@@ -1,26 +1,38 @@
 ﻿namespace ServiceControl.HeartbeatMonitoring
 {
+    using System.Collections.Generic;
     using NServiceBus;
-    using Plugin.Heartbeat.Messages;
+    using Particular.Operations.Heartbeats.Api;
     using ServiceControl.Contracts.EndpointControl;
     using ServiceControl.Contracts.Operations;
     using ServiceControl.Infrastructure.DomainEvents;
-    using ServiceControl.Monitoring;
+    using RegisterEndpointStartup = ServiceControl.Plugin.Heartbeat.Messages.RegisterEndpointStartup;
 
     class RegisterEndpointStartupHandler : IHandleMessages<RegisterEndpointStartup>
     {
-        EndpointInstanceMonitoring monitoring;
+        IEnumerable<IProcessHeartbeats> heartbeatProcessors;
         IDomainEvents domainEvents;
 
-        public RegisterEndpointStartupHandler(EndpointInstanceMonitoring monitoring, IDomainEvents domainEvents)
+        public RegisterEndpointStartupHandler(IDomainEvents domainEvents, IEnumerable<IProcessHeartbeats> heartbeatProcessors)
         {
-            this.monitoring = monitoring;
             this.domainEvents = domainEvents;
+            this.heartbeatProcessors = heartbeatProcessors;
         }
 
         public void Handle(RegisterEndpointStartup message)
         {
-            monitoring.GetOrCreateMonitor(message.Endpoint, message.Host, message.HostId, true);
+            foreach (var processor in heartbeatProcessors)
+            {
+                processor.Handle(new Particular.Operations.Heartbeats.Api.RegisterEndpointStartup
```

Comment:
Why process them synchronously?

```CSharp

Task.WhenAll(heartbeatProcessers.Select(p => p.Handle(new Particular.Operations.Heartbeats.Api.RegisterEndpointStartup {
    Endpoint = message.Endpoint,
    ...
    StartedAt = message.StartedAt
})).GetAwaiter().GetResult();
```

Reply:
We get asynchronousity on the incoming message level, but not at the heartbeatProcessors level. If each heartbeat processor handles the messages in a way that it benefits from async processing, then yes.

 Future heartbeat processors might rely heavily on cloud-based operations, where doing them in parallel would be beneficial.